### PR TITLE
Fix model/xs sampling and add test for Stepper

### LINKED ISSUE #356: Fix discrete model selection and model applicability
When sampling a discrete interaction, we sample a process using the per-process cross section calculated at the pre-step energy and then use the post-step energy to select the applicable model. Because of inconsistencies between the energy grid endpoints of the cross section tables and the model limits, it was possible for a process to be sampled but not have a model that applied at that energy. To get around this and make sure a model is always selected for the sampled process (but is this actually necessary, or can we just return an invalid model ID if none are applicable?), all our models currently have their lower applicable energy set to zero. This is wrong, but should produce correct results as long as the integral approach is used. However, without rejecting interactions based on the cross section at the post-step energy, we can end up e.g. with a stopped positron selecting bremsstrahlung because of the incorrect model energy bounds. We should figure out how to handle these inconsistencies and correctly set the model energy limits.

### PR DESCRIPTION
**Pull request by sethrj**
Prerequisites:
- [x] #424 (test harness refactor)
- [x] #426 (testing improvements for potential uninitialized memory)
- [x] #427 (testem3 and more testing)
- [ ] Refactor edge case handling

The main change to functionality here is to  recalculate post-step cross sections inside the "discrete" kernel if the particle has an energy loss process. This change should fix #356, even when integral cross section rejection is disabled. (It may be inefficient because it requires a new cross section sample and a new step, but it might be better than shoehorning the pre-step kernel into this one.)

I've also added a simple test for the `Stepper`, which is made more useful by including a new `ApplyCutoffProcess` inside the testing harness that kills particles below the cutoff in case no other absorption process is present.

### CODE REVIEW

**Comment by amandalund** [comment]
@sethrj I haven't looked at the changes yet, but I tried running the demo loop with simple-cms/testem3 and it seems to be hanging...

**Comment by sethrj** [comment]
@amandalund I added some extra debugging code while fiddling with unit tests; I can break this into multiple PRs if it makes it easier to review. I haven't yet run testem3 or simple-cms (aside from the unit tests), and we're at the point where we can actually start pulling these into the test harness now.

**Comment by amandalund** [comment]
Thanks @sethrj -- it looks like it still hangs when we don't cut it off after `max_steps`. Separating the model/cross section changes from the rest might make it a little easier, but it's also no problem for me to review as is.

**Comment by sethrj** [comment]
@amandalund I'll pull them out, PR coming right up :)

**Comment by sethrj** [comment]
By the way @amandalund : if the demo loop is hanging, you should still be able to use ctrl-C (or send SIGINT directly to the process if it's running in a wrapper) and you'll get all the output for the step iterations up to that point. That can at least tell us if it's a few tracks or all of them that are misbehaving...

**Comment by amandalund** [comment]
Ok I think since the energy loss tables don't start at zero (looks like by default the energy grid starts at 0.1 keV) with the new behavior of the `XsCalculator` low energy charged particles won't lose energy over the step, so the electrons won't die...

**Comment by sethrj** [comment]
Aha, you're right we have:
```c++
//! For now, energy loss calculation has the same behavior as cross sections
using EnergyLossCalculator = XsCalculator;
```
Welll, that makes me wonder whether the behavior was right before, since it's not like the energy loss rate is constant below 100 eV...

Thanks for finding that. Maybe I can add a template parameter or option so we can support extrapolating on the low/hi/both regions. I was planning to redo all the grid construction when tackling #253 .

Stepping plots

**Comment by amandalund** [comment]
> Welll, that makes me wonder whether the behavior was right before, since it's not like the energy loss rate is constant below 100 eV...

In the Geant4 energy loss process `AlongStepDoIt` there's also a check at the beginning if the particle is below some `lowestKinEnergy` (default 1 keV), in which case the particle is stopped and the energy is deposited locally.

They also extrapolate rather than zero outside the energy grid. Maybe they set up the lambda tables so extrapolating will produce correct results, e.g. first xs value is zero if the cross section should be zero below the first energy grid point?

**Comment by sethrj** [comment]
> In the Geant4 energy loss process `AlongStepDoIt` there's also a check at the beginning if the particle is below some `lowestKinEnergy` (default 1 keV), in which case the particle is stopped and the energy is deposited locally.

Ok, so another energy cutoff... this one is independent of material? Maybe we need to add this one to the CutoffParams and apply it during `AlongStep`.

> They also extrapolate rather than zero outside the energy grid. Maybe they set up the lambda tables so extrapolating will produce correct results, e.g. first xs value is zero if the cross section should be zero below the first energy grid point?

That might work for nonuniform energy grids where you can create a discontinuity, but from what I can tell everything but "inverse range" is a uniform grid: so if your lowest point is zero there's always a nontrivial decrease between the lowest positive value and the zero.

We can do better/different, of course. But perhaps the best option for the time being is to deposit all energy if the track is below the lowest energy grid point.

**Comment by sethrj** [comment]
@amandalund I worked around the energy loss extrapolation bug, and additionally added tests for below-the-grid energy loss *and* added a unit test with Geant4 integration for TestEm3.

**Comment by sethrj** [comment]
Also @amandalund  if it would help, I think I can break out some of the new tests and the Geant4 changes into another separate PR 😂

File: src/celeritas/grid/XsCalculator.hh [hunk 1]
```diff
@@ -83,21 +86,33 @@ XsCalculator::XsCalculator(const XsGridData& grid, const Values& values)
...
+ * bounds? (And is it OK if those endpoints are a bit fuzzy because of the
+ * logarithms?)
  */
 CELER_FUNCTION real_type XsCalculator::operator()(Energy energy) const
 {
     const UniformGrid loge_grid(data_.log_energy);
     const real_type   loge = std::log(energy.value());
 
-    // Snap out-of-bounds values to closest grid points
+    // Cross section is zero outside of defined grid
```
Comment (amandalund) [inline src/celeritas/grid/XsCalculator.hh:99]: Now I'm not totally convinced this is the right behavior. Since Geant4 extrapolates rather than zeros outside the grid, maybe as long as we are importing our physics tables from Geant4 we should do the same thing? Looking at how they build the lambda tables, they use a "start from null" flag to indicate whether the first cross section value on the table should be zero (presumably so that extrapolating will produce the correct result below the threshold for that process) -- however, it's not true for all processes. Also, should the cross section really be zero above the energy grid? It's a little awkward too to change the behavior of the eloss calculator so that below the grid the dE/dx is zero rather than the first value and then hack in a way to set it back to the first value.

**Review by amandalund — COMMENTED** [review]
haha, only if you really want to ;). I kind of broke it up on my own -- here's a couple questions on the model/xs sampling while I look over the rest..

File: test/corecel/data/testem3-flat.org.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# -*- coding: utf-8 -*-
+# Copyright 2022 UT-Battelle, LLC and other Celeritas Developers.
+# See the top-level COPYRIGHT file for details.
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""
+ORANGE testem3 input definition using python.
```
[the remark this answers is not in the document]
Comment (sethrj) [reply test/corecel/data/testem3-flat.org.py:6]: whoops, these were intended to be in `celeritas/data`... thanks for finding the duplicates 😳

File: src/celeritas/phys/PhysicsStepUtils.hh [hunk 3]
```diff
@@ -313,31 +358,14 @@ select_discrete_interaction(const ParticleTrackView& particle,
...
-        // step \f$ \sigma_{\max} \f$.
-        if (generate_canonical(rng) * pstep.per_process_xs(ppid) > xs)
-        {
-            // No interaction occurs; reset the physics state and continue
-            // tracking
-            return physics.scalars().integral_rejection_action();
-        }
+        // Particle energy has changed over the step and the interaction is no
+        // longer valid
+        return physics.scalars().integral_rejection_action();
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/phys/PhysicsStepUtils.hh:368]: Yes, that's correct. It's not optimal: we should be constructing the model limits based on the process cross sections we're given (or the other way around if we generate cross sections ourselves rather than import them from geant4). But I didn't want to open that can of worms.

File: src/celeritas/grid/XsCalculator.hh (hunk 1 above)
Comment (sethrj) [reply src/celeritas/grid/XsCalculator.hh:99]: ok, maybe @stognini we also should add that flag to the export data. Is it exposed as an accessor?

Lol yes it is awkward. Let me split off the changes that *don't* require touching the model/cross section extrapolation, and then I can improve what we're doign here

Comment (amandalund) [reply src/celeritas/grid/XsCalculator.hh:99]: > Is it exposed as an accessor?

Hmm no. It also only apples if the minimum primary energy for the process is greater than the minimum kinetic energy of the tables. But I guess whatever the first xs value is is what should be used below the energy grid.

Comment (stognini) [reply src/celeritas/grid/XsCalculator.hh:99]: As @amandalund  said, the `startFromNull` flags in both `G4VEmProcess::BuildLambdaTable()` and `G4VEenergyLossProcess::BuildLambdaTable()` are local variables. To make things worse, we can't use the `MinPrimaryEnergy(...)` to replicate the criterion just mentioned by @amandalund at import time, because these are a protected members of each class. So it is not (currently) obvious how we could get these flag values out, but I'll dig more to try and find a workaround without any changes to G4.

Also, not directly related to this topic, but might be relevant to the conversation: G4 does a similar thing to micro-xs. See comment here:

https://github.com/celeritas-project/celeritas/blob/master/src/celeritas/ext/detail/ImportProcessConverter.cc#L618-L641

**Comment by sethrj** [comment]
OK @amandalund now at least we've got *some* testing of the stepper on the main branch, I think it's OK to put this branch on hold... I've got some other tasks to accomplish in the next couple of days anyway. Thanks for the feedback so far though.

File: src/celeritas/grid/XsCalculator.hh (hunk 1 above)
Comment (sethrj) [reply src/celeritas/grid/XsCalculator.hh:99]: @stognini If I'm understanding you and your comment correctly, do we need to have flags for whether to set off-the-end values to zero? If "Geant4 simply uses the next/previous bin value when the vector's front/back are zero" when inside the energy range, and if it uses the actual first and last value outside the energy range, then the current implementation in celeritas will give a different (nonzero) value when accessing out-of-range energies.

Comment (stognini) [reply src/celeritas/grid/XsCalculator.hh:99]: If we don't do exactly the same as Geant4, then it is useful to know about any forced rewrite in the physics vectors: Either a zero added at the front of a macro lambda table, or repeated non-zero values at the front/back of micro-xs tables.

Did we reach a consensus on how we want to deal with process limits, as discussed in #356 ?

**Comment by sethrj** [comment]
Did I say days? I "meant" months. We can probably revisit this when we refactor the physics data.
